pkg/object/mqttproxy/client.go

```diff
@@ -257,9 +244,40 @@ func (c *Client) processPuback(puback *packets.PubackPacket) {
 	c.session.puback(puback)
 }
 
+// runPipeline will run MQTT pipeline by using packet.
+// it will return an error if MQTT pipline set MQTTContext to Disconnect or Drop.
+func (c *Client) runPipeline(packet packets.ControlPacket) error {
+	if c.broker.pipeline == "" {
+		return nil
+	}
+
+	pipe, err := pipeline.GetPipeline(c.broker.pipeline, context.MQTT)
+	if err != nil {
+		logger.SpanErrorf(nil, "get pipeline %v failed, %v", c.broker.pipeline, err)
+		return nil
+	}
+
+	ctx := context.NewMQTTContext(stdcontext.Background(), c.broker.backend, c, packet)
+	pipe.HandleMQTT(ctx)
+	if ctx.Disconnect() {
+		c.close()
+		return errors.New("pipeline set disconnect")
+	}
+	if ctx.Drop() {
+		return errors.New("pipeline set drop")
+	}
+	return nil
+}
+
 func (c *Client) processSubscribe(packet *packets.SubscribePacket) {
 	logger.SpanDebugf(nil, "client %s subscribe %v with qos %v", c.info.cid, packet.Topics, packet.Qoss)
-	err := c.broker.topicMgr.subscribe(packet.Topics, packet.Qoss, c.info.cid)
+	err := c.runPipeline(packet)
+	if err != nil {
+		logger.SpanDebugf(nil, "client %v process subscribe %v failed, %v", c.info.cid, packet.Topics, err)
+		return
+	}
+
```

Comment:
Duplicated codes in each packet routines, according to DRY, it's very great if you refactor it”

Reply:
... Not all packets run pipeline... if we received some packets, it's an error, like `Connect` about connect established, `Connack`, `Suback`, `Unsuback`; some packet we do nothing to it, like `Pingresp`; some packet we just send response like `Pingreq`... Is this want you mean?

Reply:
updated!